Navigation API is making O(# entries in back forward list) sync IPC calls each time a document is created

RupinMittal opened the pull request:
#### a836f7d0d68af50c503f55fa0358e6bed9d99fe1
<pre>
Navigation API is making O(# entries in back forward list) sync IPC calls each time a document is created
<a href="https://bugs.webkit.org/show_bug.cgi?id=297571">https://bugs.webkit.org/show_bug.cgi?id=297571</a>
<a href="https://rdar.apple.com/158648936">rdar://158648936</a>

Reviewed by Charlie Wolfe.

When a new navigation object is created, we need to initialize
it&apos;s history entries. Back in the initial implementation of the
Navigation API in <a href="https://commits.webkit.org/273532@main">https://commits.webkit.org/273532@main</a>,
this was done by getting each individual entry by calling
WebBackForwardListProxy::itemAtIndex(). This gets the entry
by making a sync IPC to the UIProcess.

This means every time we make a new navigation object (which
is every time a new document is created), we are making a sync
IPC call for each individual entry in the back forward list.

This is not great for performance.

We fix this by adding a new IPC to allow us to get the whole list at
once. From testing, this reduces the PLT regression caused by enabling
the Navigation API by almost 2% (16.2% --&gt; 14.6%).

To fully eliminate the remaining 15% regression, we&apos;ll need to
completely eliminate this IPC or make it async. I&apos;m not yet sure
how to remove this call or make it async since a second navigation
happening before the async call finishing would mean the list of
entries is corrupted. Still, this is useful since a single IPC is
better than many. It reduces the regression a bit and allItems()
is called other places as well.

No new tests as this doesn&apos;t change user-facing behavior, just
improves performance.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):

Instead of first making a sync IPC to get the backCount and then
making a sync IPC via itemForIndex() to get each item, call the
new WebBackForwardListProxy::allItems() function which will get
all the items with a single sync IPC call.

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebKit/UIProcess/WebBackForwardList.h:
(WebKit::WebBackForwardList::allItems const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::backForwardAllItems):

New function that returns all the items in the back forward list.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::allItems):

This will make the IPC from the Web Process to the UI Process
to get all the items from WebPageProxy::backForwardAllItems().

* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKitLegacy/mac/History/BackForwardList.h:

Canonical link: <a href="https://commits.webkit.org/298922@main">https://commits.webkit.org/298922@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7930296d9ff8e9f4b4a41899415177b09f12416e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117106 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123173 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69106 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/004977ca-26d4-487a-b3d8-c6c65957b8fa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118980 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37471 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45361 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88895 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7181e638-f462-4700-ae0c-f09fc8bbdd0e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120039 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29858 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105029 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69379 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28918 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23139 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66828 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99248 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23299 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126315 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43998 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33054 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97572 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97371 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42712 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20641 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40379 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18700 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43873 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49485 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43340 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46686 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45050 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/history/BackForwardClient.h:
```diff
@@ -30,13 +30,12 @@
 #include <WebCore/BackForwardFrameItemIdentifier.h>
 #include <WebCore/BackForwardItemIdentifier.h>
 #include <WebCore/FrameIdentifier.h>
+#include <WebCore/HistoryItem.h>
```
I doubt this is needed. Just keep the forward declaration.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
```diff
@@ -85,6 +85,25 @@ void WebBackForwardListProxy::goToItem(HistoryItem& item)
+
+    Vector<Ref<HistoryItem>> allItems;
+    for (auto frameState : allFrameStates)
+        allItems.append(toHistoryItem(historyItemClient, frameState));
```
```suggestion
    for (Ref frameState : allFrameStates)
        allItems.append(toHistoryItem(historyItemClient, WTFMove(frameState)));
```

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10039,6 +10039,24 @@ void WebPageProxy::backForwardGoToItemShared(BackForwardItemIdentifier itemID, C
+{
+    Vector<Ref<FrameState>> allItems;
+
+    for (auto item : m_backForwardList->allItems()) {
```
```suggestion
    for (Ref item : m_backForwardList->allItems()) {
```

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10039,6 +10039,24 @@ void WebPageProxy::backForwardGoToItemShared(BackForwardItemIdentifier itemID, C
+        else
+            frameState = item->mainFrameState();
+
+        allItems.append(*frameState);
```
```suggestion
        allItems.append(frameState.releaseNonNull());
```

RupinMittal committed:
Navigation API is making O(# entries in back forward list) sync IPC calls each time a document is created
https://bugs.webkit.org/show_bug.cgi?id=297571
rdar://158648936

Reviewed by Charlie Wolfe.

When a new navigation object is created, we need to initialize
it's history entries. Back in the initial implementation of the
Navigation API in https://commits.webkit.org/273532@main,
this was done by getting each individual entry by calling
WebBackForwardListProxy::itemAtIndex(). This gets the entry
by making a sync IPC to the UIProcess.

This means every time we make a new navigation object (which
is every time a new document is created), we are making a sync
IPC call for each individual entry in the back forward list.

This is not great for performance.

We fix this by adding a new IPC to allow us to get the whole list at
once. From testing, this reduces the PLT regression caused by enabling
the Navigation API by almost 2% (16.2% --> 14.6%).

To fully eliminate the remaining 15% regression, we'll need to
completely eliminate this IPC or make it async. I'm not yet sure
how to remove this call or make it async since a second navigation
happening before the async call finishing would mean the list of
entries is corrupted. Still, this is useful since a single IPC is
better than many. It reduces the regression a bit and allItems()
is called other places as well.

No new tests as this doesn't change user-facing behavior, just
improves performance.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):

Instead of first making a sync IPC to get the backCount and then
making a sync IPC via itemForIndex() to get each item, call the
new WebBackForwardListProxy::allItems() function which will get
all the items with a single sync IPC call.

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebKit/UIProcess/WebBackForwardList.h:
(WebKit::WebBackForwardList::allItems const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::backForwardAllItems):

New function that returns all the items in the back forward list.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::allItems):

This will make the IPC from the Web Process to the UI Process
to get all the items from WebPageProxy::backForwardAllItems().

* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKitLegacy/mac/History/BackForwardList.h:

Canonical link: https://commits.webkit.org/298922@main
